app/models/spree/variant.rb

```diff
@@ -297,5 +298,11 @@ def destruction
       exchange_variants(:reload).destroy_all
       yield
     end
+
+    def ensure_unit_value
+      return unless product&.variant_unit == "items" && unit_value.nil?
+
+      self.unit_value = 1.0
+    end
```

Comment:
I agree this is a quick patch that may solve the issue but it feels like we're pilling up callbacks onto callbacks which make our models slow and hard to reason about. IMO it'd a much maintainable approach to create a `ChangeProductUnit` and explicitly call it from the controller in charge. This way the variant doesn't pay the toll of yet another callback in every single INSERT/UPDATE to DB.

Reply:
I think the root cause of this is that the `Product` and `Variant` models are not separated very well, and that the UI reinforces this commingling. So if we create a separate service and put the responsibility on the controller in charge, then the messiness spreads out to that controller as well. Then if another controller, say an API controller, is created, we have to ask "does this new controller need to use this service?" rather than just keeping the weirdness limited to these two classes. 

I definitely agree that it looks like a hack upon a hack, but in this specific instance I'd lean toward the callback approach instead of a service.

Reply:
I'm not 100% sold on that but it's fair enough :+1:

Reply:
Agreed it's not great, but these problems are piling up all the time. 

We're setting a value on the object during the `before_validation` step, so it's not really an extra strain on the DB in terms of transactions. Eg: we're not triggering additional database updates inside of an `after_save` callback for example, which would be a lot worse...